### ISSUE TITLE
@parcel/bundler-default: The expression evaluated to a falsy value (CSS modules and non-modules)

### ISSUE THREAD

**Issue by nartallax** [parcel-bundler/parcel#8716]
# 🐛 bug report

In particular circumstances I get following error when building my project:

```bash
@parcel/bundler-default: The expression evaluated to a falsy value:

  (0, _assert().default)(bundle !== 'root' && bundle != null)
```

## 🎛 Configuration (.babelrc, package.json, cli command)

[Minimal example is here](https://github.com/nartallax/parcel-css-bug)

Key points: if there is non-module CSS file and more than one module CSS files in project - build will fail.

Build command is

```bash
./node_modules/.bin/parcel build src/index.html --no-cache
```

## 🤔 Expected Behavior

Build should not fail. I don't think this CSS structure is wrong or unobvious; at least it should lead to better error than just "assertion failed"

## 😯 Current Behavior

Build fails with following stack trace:
```bash
@parcel/bundler-default: The expression evaluated to a falsy value:

  (0, _assert().default)(bundle !== 'root' && bundle != null)


  AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

    (0, _assert().default)(bundle !== 'root' && bundle != null)

      at getBundleFromBundleRoot (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/bundler-default/lib/DefaultBundler.js:1018:27)
      at createIdealGraph (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/bundler-default/lib/DefaultBundler.js:692:21)
      at Object.bundle (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/bundler-default/lib/DefaultBundler.js:122:19)
      at BundlerRunner.bundle (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:288:23)
      at async Object.run (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:156:17)
      at async RequestTracker.runRequest (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/RequestTracker.js:756:20)
      at async Object.run (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/requests/ParcelBuildRequest.js:56:7)
      at async RequestTracker.runRequest (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/RequestTracker.js:756:20)
      at async Parcel._build (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/Parcel.js:400:11)
      at async Parcel.run (/Users/MY_USERNAME/parcel_css_bug/node_modules/@parcel/core/lib/Parcel.js:276:18)
```      

## 🔦 Context

I'm just developing some websites lol

## 💻 Code Sample

[Minimal example is here](https://github.com/nartallax/parcel-css-bug)

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.8.2 |
| Node             | v16.16.0 |
| npm/Yarn         | 8.11.0 |
| Operating System | macOS 13.0.1 |

**Comment by mischnic** [comment]
cc @AGawrys

**Comment by oe** [comment]
same here with parcel@2.8.2

**Comment by JulianGibelli** [comment]
Same here! parcel version 2.8.2
![image](https://user-images.githubusercontent.com/79370118/210888225-6ec31e9c-6f3a-4d14-955d-69b31c0c928f.png)

**Comment by littleLane** [comment]
+1

**Comment by devongovett** [comment]
Should be fixed by #8762

**Comment by shepmaster** [comment]
I see that this is closed and should have been fixed by 2.8.3. However, after upgrading from 2.8.0 to 2.8.3, I still see the same error:

```
🚨 Build failed.

@parcel/bundler-default: The expression evaluated to a falsy value:

  (0, _assert().default)(bundle !== 'root' && bundle != null)


  AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

    (0, _assert().default)(bundle !== 'root' && bundle != null)

      at getBundleFromBundleRoot (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:1033:27)
      at createIdealGraph (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:698:21)
      at Object.bundle (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:122:19)
      at BundlerRunner.bundle (...project.../node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:288:23)
      at async Object.run (...project.../node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:156:17)
      at async RequestTracker.runRequest (...project.../node_modules/@parcel/core/lib/RequestTracker.js:756:20)
      at async Object.run (...project.../node_modules/@parcel/core/lib/requests/ParcelBuildRequest.js:56:7)
      at async RequestTracker.runRequest (...project.../node_modules/@parcel/core/lib/RequestTracker.js:756:20)
      at async Parcel._build (...project.../node_modules/@parcel/core/lib/Parcel.js:400:11)
at async Parcel._startNextBuild (...project.../node_modules/@parcel/core/lib/Parcel.js:298:24)
```

What should I look for in my project to see what might be unique from the cases that were already fixed?

**Comment by AGawrys** [comment]
Might just be the same symptom. Do you have a repro or repo I could look at? @shepmaster

**Comment by shepmaster** [comment]
Here is a minimized example: [repro.zip](https://github.com/parcel-bundler/parcel/files/10503120/repro.zip). Unzip it, run `yarn && yarn build` to see the error. I'm using Node v19.4.0 on macOS.

My gut tells me that it has to do with using the `extends` feature of CSS modules, but I've got nothing concrete to back that up. Thanks!

**Comment by itaylor** [comment]
I also have this problem, it's definitely caused by css modules `composes`, which my project makes extensive use of. 

It looks like the file referenced in the `compose` statement is added as an asset dependency, but then doing that messes up the bundleRootGraph somehow in `createIdealGraph` in `DefaultBundler` (perhaps only if the composed file is also a css module?).  

Using @shepmaster's repro from above and then removing the statement:
```js
             asset.addDependency({
                specifier: ref.specifier,
                specifierType: 'esm',
              });
 ``` 
 From the `CSSTransformer`'s handling of composes solves the error in the build, and it still retains the composed css.  

It looks like this was added fairly recently to improve css import ordering:
https://github.com/parcel-bundler/parcel/commit/ddae31a917c9482d461871958335e34c755cf8ce#diff-4a76ab5816b5309c818b936e2b32e9beeca804ecf94c2bd18cc22108e02b5274

**Comment by devongovett** [comment]
Should be fixed by #8885

**Comment by martindavid** [comment]
@devongovett how to get the fixed version for this? Should there be a patch version release for this?

**Comment by chrissound** [comment]
I got this error when running `parcel` with additional html files (with 0 css links/references) in a `article/` directory.

Basically `parcel webroot/index.html` works, while `parcel webroot/index.html webroot/article/*.html` resulted in this error...

**Comment by mischnic** [comment]
There will be a release including the fix very soon

**Comment by geremsadaimary** [comment]
why is this closed without being fixed

**Comment by anupxroutray** [comment]
+1 @geremsadaimary @mischnic

**Comment by sohamshet26** [comment]
> I see that this is closed and should have been fixed by 2.8.3. However, after upgrading from 2.8.0 to 2.8.3, I still see the same error:
> 
> ```
> 🚨 Build failed.
> 
> @parcel/bundler-default: The expression evaluated to a falsy value:
> 
>   (0, _assert().default)(bundle !== 'root' && bundle != null)
> 
> 
>   AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:
> 
>     (0, _assert().default)(bundle !== 'root' && bundle != null)
> 
>       at getBundleFromBundleRoot (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:1033:27)
>       at createIdealGraph (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:698:21)
>       at Object.bundle (...project.../node_modules/@parcel/bundler-default/lib/DefaultBundler.js:122:19)
>       at BundlerRunner.bundle (...project.../node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:288:23)
>       at async Object.run (...project.../node_modules/@parcel/core/lib/requests/BundleGraphRequest.js:156:17)
>       at async RequestTracker.runRequest (...project.../node_modules/@parcel/core/lib/RequestTracker.js:756:20)
>       at async Object.run (...project.../node_modules/@parcel/core/lib/requests/ParcelBuildRequest.js:56:7)
>       at async RequestTracker.runRequest (...project.../node_modules/@parcel/core/lib/RequestTracker.js:756:20)
>       at async Parcel._build (...project.../node_modules/@parcel/core/lib/Parcel.js:400:11)
> at async Parcel._startNextBuild (...project.../node_modules/@parcel/core/lib/Parcel.js:298:24)
> ```
> 
> What should I look for in my project to see what might be unique from the cases that were already fixed?

i have also got this type of error but i just cleared my parcel cashe and the error was resolved
